### PR TITLE
TF-934 Support edit route, open new tab email on browser

### PR DESCRIPTION
**Pull request by dab246**
#934 

Current supported browser address path parameters: `https:domain.com/id?type=&context=&q=`

- `type`: Distinguish search and non-search status. Get 2 values `normal` and `search`
- `context`: Get the value of `mailboxId`
- `/id?`: Get  the value of `emailId`
- `q`: Get simple search keywords

### Resolved

- Open in New Tab

https://user-images.githubusercontent.com/80730648/201632425-5f6f2bcf-7791-4e57-bdeb-5bebd3e1888b.mov

- Change router

https://user-images.githubusercontent.com/80730648/201188569-7dfdd959-a73f-439f-a7eb-41f48dad4411.mp4

- Page not found

https://user-images.githubusercontent.com/80730648/201187135-9e55c42d-4eb7-4144-a2d9-991cfdac8397.mov

### CODE REVIEW

**Review by hoangdat — APPROVED** [review]

File: lib/main/routes/route_utils.dart [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+import 'package:core/utils/app_logger.dart';
+import 'package:jmap_dart_client/jmap/core/id.dart';
+import 'package:jmap_dart_client/jmap/mail/email/email.dart';
+import 'package:jmap_dart_client/jmap/mail/mailbox/mailbox.dart';
+import 'package:tmail_ui_user/features/login/data/extensions/service_path_extension.dart';
+import 'package:tmail_ui_user/features/thread/domain/model/search_query.dart';
+import 'package:tmail_ui_user/main/routes/navigation_router.dart';
+import 'package:universal_html/html.dart' as html;
+
+abstract class RouteUtils {
```
Comment (sherlockvn) [inline lib/main/routes/route_utils.dart:14]: why RouteUtils is abstract ?

File: model/lib/email/presentation_email.dart [hunk 2]
```diff
@@ -84,6 +89,10 @@ class PresentationEmail with EquatableMixin {
     sentAt,
     replyTo,
     preview,
-    hasAttachment
+    hasAttachment,
+    mailboxIds,
```
Comment (sherlockvn) [inline model/lib/email/presentation_email.dart:93]: why does props need to have these fields, IMO, i think that routeWeb contains mailboxId. And mailboxId is unique, so i think we don't need mailboxNames

**Comment by chibenwa** [comment]
Looks awesome!

(This might be another ticket but...) Is it possible to right click on a mailbox (right side panel) to open it into a new TAB ? cc @hoangdat

File: lib/features/thread/data/datasource_impl/thread_datasource_impl.dart [hunk 3]
```diff
@@ -92,4 +94,14 @@ class ThreadDataSourceImpl extends ThreadDataSource {
       _exceptionThrower.throwException(error);
     });
   }
+
+  @override
+  Future<PresentationEmail> getEmailById(AccountId accountId, EmailId emailId, {Properties? properties}) {
+    return Future.sync(() async {
```
[the remark this answers is not in the document]
Comment (dab246) [reply lib/features/thread/data/datasource_impl/thread_datasource_impl.dart:100]: We need to handle synchronous statements in this datasource.

File: lib/features/thread/data/network/thread_api.dart [hunk 4]
```diff
@@ -139,4 +139,29 @@ class ThreadAPI {
...
+    final result = await (jmapRequestBuilder
+        ..usings(getEmailMethod.requiredCapabilities))
+      .build()
+      .execute();
+
+    final resultList = result.parse<GetEmailResponse>(
+      getEmailInvocation.methodCallId,
+      GetEmailResponse.deserialize);
+
+    return resultList!.list.first;
```
[the remark this answers is not in the document]
Comment (dab246) [reply lib/features/thread/data/network/thread_api.dart:165]: In this field we don't need to care about the error so we just `try/catch` it to ignore it when it failed.

File: lib/main/routes/route_utils.dart (hunk 1 above)
Comment (dab246) [reply lib/main/routes/route_utils.dart:14]: We don't want to create a constructor for this class.

File: model/lib/email/presentation_email.dart (hunk 2 above)
Comment (dab246) [reply model/lib/email/presentation_email.dart:93]: We need `mailboxNames` to display `label` for each email when searching

**Comment by dab246** [comment]
> Looks awesome!
> 
> (This might be another ticket but...) Is it possible to right click on a mailbox (right side panel) to open it into a new TAB ? cc @hoangdat

OK, I think it's pretty simple, I'll create a ticket and implement it in the next sprint. Because this PR has quite a lot of `changes` already

**Comment by chibenwa** [comment]
> (This might be another ticket but...) Is it possible to right click on a mailbox (right side panel) to open it into a new TAB ? cc @hoangdat

https://github.com/linagora/tmail-flutter/issues/1201

**Review by hoangdat — APPROVED** [review]

**Pull request merged**